src/wallet/wallet2.cpp

```diff
@@ -1759,8 +1759,19 @@ void wallet2::update_pool_state(bool refreshed)
 void wallet2::fast_refresh(uint64_t stop_height, uint64_t &blocks_start_height, std::list<crypto::hash> &short_chain_history)
 {
   std::list<crypto::hash> hashes;
-  size_t current_index = m_blockchain.size();
 
+  const uint64_t checkpoint_height = m_checkpoints.get_max_height();
+  if (stop_height > checkpoint_height && m_blockchain.size()-1 < checkpoint_height)
+  {
+    // we will drop all these, so don't bother getting them
+    uint64_t missing_blocks = m_checkpoints.get_max_height() - m_blockchain.size();
+    while (missing_blocks-- > 0)
+      m_blockchain.push_back(crypto::null_hash); // maybe a bit suboptimal, but deque won't do huge reallocs like vector
+    m_blockchain.push_back(m_checkpoints.get_points().at(checkpoint_height));
+    m_local_bc_height = m_blockchain.size();
+  }
+
+  size_t current_index = m_blockchain.size();
   while(m_run.load(std::memory_order_relaxed) && current_index < stop_height)
   {
     pull_hashes(0, blocks_start_height, short_chain_history, hashes);
```

Comment:
I think first arg here should be the same as m_blockchain.size()? Otherwise hashes are pulled from block zero and the fast refresh breaks below (`if(current_index >= m_blockchain.size())`).

Haven't tested this, but I noticed that fast refresh never continued with this patch.

Reply:
Can you give a step by step repro case ?

Reply:
- Create a new wallet that uses the approximate height calculation (i.e. without being connected to daemon). 
- Connect wallet to daemon and start syncing last months of blocks. 
- With log level 2 you'll see "skipped block by timestamp" instead of "skipped block by height" when syncing the blocks from last checkpoint (1390000) to wallets creation height (14xxxxx) which indicates that that fast_refresh isn't used for those blocks.

Reply:
hold your horses. I might have been fooled by #2724. Last checkpoint is not one month old. I'll try to reproduce this with cli instead later today.

Reply:
OK. Did some more testing. I could reproduce this even with #2724 applied. The calculated block height 1 month back is 1410159 and this is what i get in the log with a fresh wallet:
``wallet.wallet2	src/wallet/wallet2.cpp:1264	Skipped block by timestamp, height: 1390100, block time 1504343216, account time 1509381622
2017-10-30 16:40:36.568	  0x70000ff6b000	DEBUG	wallet.wallet2	src/wallet/wallet``

Reply:
Can you put that in a new bug so it's not forgotten ? I'll look at that when I'm done with what I'm doing now.

Reply:
Just to be clear. The bug is only related to this PR. Not reproducible without it. I suggest it's fixed before this is merged. I can still put the report in a new issue if you prefer that?

Reply:
No, if that's broken by this patch, then it should stay here. You were talking about a line outside the patch so it was not clear :)

Reply:
Fixed